# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Kelly Simmons
#### Github: https://github.com/Kells2234/JSON-Sentiment-M4

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [2]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package            Version
------------------ ---------
asttokens          2.2.1
backcall           0.2.0
blis               0.7.10
catalogue          2.0.9
certifi            2023.7.22
charset-normalizer 3.2.0
click              8.1.6
colorama           0.4.6
comm               0.1.3
confection         0.1.0
cymem              2.0.7
debugpy            1.6.7
decorator          5.1.1
en-core-web-sm     3.6.0
executing          1.2.0
idna               3.4
ipykernel          6.25.0
ipython            8.14.0
jedi               0.18.2
Jinja2             3.1.2
joblib             1.3.1
jupyter_client     8.3.0
jupyter_core       5.3.1
langcodes          3.3.0
MarkupSafe         2.1.3
matplotlib-inline  0.1.6
murmurhash         1.0.9
nest-asyncio       1.5.6
nltk               3.8.1
numpy              1.25.1
packaging          23.1
parso              0.8.3
pathy              0.10.2
pickleshare        0.7.5
pip                23.2.1
platformdirs       3.9.1
preshed      

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [2]:
import requests
import json

# result = json.loads(requests.get('https://api.lyrics.ovh/v1/They Might Be Giants/Birdhouse in your soul').text)

# Setting up Lyrics Genius
import lyricsgenius
genius = lyricsgenius.Genius("7V8yngWGXXqB8-GgHm6Ts7wD8YcM5QU4Nw8ZzzaNxDb9b_4dZlUxIi5JY9iZ_sAL")

# Setting the artist variable to the corrrect artist
artist = genius.search_artist("They Might Be Giants", max_songs=5, sort="title")

# Find the "Birdhouse in Your Soul" song
song = artist.song('Birdhouse In Your Soul')

# Saving the Lyrics
lyrics = song.lyrics

# Creating the dictionary with song information
song_dict = {
    'Artist': 'They Might Be Giants',
    'Song': 'Birdhouse In Your Soul',
    'Lyrics': lyrics
}

with open('birdhouse_lyrics.json', 'w') as new_file:
    json.dump(song_dict, new_file)

Searching for songs by They Might Be Giants...

Song 1: "200 Sbemails (for Homestar Runner)"
Song 2: "2082"
Song 3: "25 O’Clock"
Song 4: "32 Footsteps"
Song 5: "3rd Amendment"

Reached user-specified song limit (5).
Done. Found 5 songs.
Searching for "Birdhouse In Your Soul" by They Might Be Giants...
Done.


2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [3]:
with open ('birdhouse_lyrics.json', 'r') as file:
    dictionary = json.load(file)

dictionary['Lyrics']

"52 ContributorsBirdhouse in Your Soul Lyrics[Bridge]\nI'm your only friend\nI'm not your only friend\nBut I'm a little glowing friend\nBut really I'm not actually your friend\nBut I am\n\n[Chorus]\nBlue canary in the outlet by the light switch\nWho watches over you\nMake a little birdhouse in your soul\nNot to put too fine a point on it\nSay I'm the only bee in your bonnet\nMake a little birdhouse in your soul\n\n[Verse 1]\nI have a secret to tell\nFrom my electrical well\nIt's a simple message and I'm\nLeaving out the whistles and bells\nSo the room must listen to me\nFilibuster vigilantly\nMy name is blue canary\nOne note spelled l-i-t-e\nMy story's infinite\nLike the Longines Symphonette\nIt doesn't rest\nSee They Might Be Giants LiveGet tickets as low as $58You might also like[Chorus]\nBlue canary in the outlet by the light switch\nWho watches over you\nMake a little birdhouse in your soul\nNot to put too fine a point on it\nSay I'm the only bee in your bonnet\nMake a little birdh

In [6]:
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

with open('lyrics.json', 'r') as file:
    data = json.load(file)

lyrics = data["lyrics"]



print("Lyrics:")
print(lyrics)

nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("spacytextblob")
doc = nlp(lyrics)

polarity_score = doc._.polarity

print("Polarity Score:", polarity_score)

Lyrics:
Lyrics of the song
Polarity Score: 0.0


3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [7]:
def write_song_from_api_to_JSON(artist, song, filename):
    genius = lyricsgenius.Genius("7V8yngWGXXqB8-GgHm6Ts7wD8YcM5QU4Nw8ZzzaNxDb9b_4dZlUxIi5JY9iZ_sAL")
    artist = genius.search_artist(artist, max_songs=1, sort="title")
    song = artist.song(song)
    with open(filename, 'w') as filename:
        json.dump(song.lyrics, filename)  

write_song_from_api_to_JSON("Prince", "The Beautiful Ones", "the_beautiful_ones_lyrics.json")
write_song_from_api_to_JSON("The Gap Band", "Wednesday Lover", "wednesday_lover_lyrics.json")
write_song_from_api_to_JSON("Ella Mai", "Trip", "trip_lyrics.json")
write_song_from_api_to_JSON("Vividry", "Your Good Lies", "your_good_lies_lyrics.json")


Searching for songs by Prince...

Changing artist name to 'RZA'
Song 1: "."

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "The Beautiful Ones" by RZA...
Done.
Searching for songs by The Gap Band...

Song 1: "All of My Love"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "Wednesday Lover" by The Gap Band...
Done.
Searching for songs by Ella Mai...

Song 1: "10,000 Hours"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "Trip" by Ella Mai...
Done.
Searching for songs by Vividry...

Song 1: "A Million Signs"

Reached user-specified song limit (1).
Done. Found 1 songs.
Searching for "Your Good Lies" by Vividry...
Done.


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [13]:
def polarity_score_from_lyrics(song, filename):
    nlp = spacy.load('en_core_web_sm')
    nlp.add_pipe('spacytextblob')
    with open(filename, 'r') as f:
        lyrics = json.load(f)
    doc = nlp(lyrics)
    polarity = doc._.blob.polarity
    print(f'{song} Polarity: {polarity}')

polarity_score_from_lyrics("The Beautiful Ones", "the_beautiful_ones_lyrics.json")
polarity_score_from_lyrics("Wednesday Lover", "wednesday_lover_lyrics.json")
polarity_score_from_lyrics("Trip", "trip_lyrics.json")
polarity_score_from_lyrics("Your Good Lies", "your_good_lies_lyrics.json")

The Beautiful Ones Polarity: 0.30047790750915754
Wednesday Lover Polarity: 0.28204545454545454
Trip Polarity: -0.0761261261261261
Your Good Lies Polarity: 0.7


The Beautiful Ones (0.30): I was thinking that the program would give a negative sentiment for this song. However I had to rethink that because the program does not take into account the actual music of the song. Just going off the words, I understand the scoring.

Wednesday Lover (0.28): This is a pure love song I thought the score would be higher.

Trip (-0.07): This is also is a love song. However, looking at the lyrics I can see why the song got the score it did. The artist uses lyrics to bring her feelings to life but the program may see them as negative.

Your Good Lies (0.7): This song got the highest score of them all which is surprising. However reading through the lyrics, I see why it got the score it did.